### ISSUE TITLE
Project analysis code to enable automating PyPI project analysis with FawltyDeps

### ISSUE THREAD

**Issue by zz1874** [tweag/FawltyDeps#376]
We aim to automate the analysis of popular PyPI projects using FawltyDeps. Manual analysis of the 10 PyPI projects has revealed several aspects that need automated detection, such as conditional imports, dynamic imports, and other alternative imports. We are also interested in dependency declaration formats and project code directory structures.

This plan outlines the tasks and subtasks required to automate the analysis process efficiently:
### 1. Initialization:
- [ ] Add relevant PyPI projects from manual analysis to the real projects tests suite.
- [x] Create another branch (or another Git repository) for this automating analysis.

### Project Source Code Analysis
Module that finds different types of problematic imports and outputs them in a single JSON or a dict-like data structure.
The detection should be able to flag cases where more than one import category applies to the code, e.g. it's both dynamic and conditional.

Whenever it's easy (like if the import name is easily found in an AST field) record the module or package name.

#### Conditional Import Detection
- [x] #377
   
Write code to parse Python source files (reuse code from core FD) and Identify conditional imports using AST or regular expressions:
```python
try:
    import foo
except ImportError:
    pass
```

#### Alternative Import Detection:
- [x] #378
```python
try:
    import foo
except ImportError:
    import bar
```
#### Dynamic Import Detection:
- [x] #379
```python
try:
    module = importlib.import_module("foo")
except ImportError:
    pass
```
or
```python
module = importlib.import_module("foo")
```
or
```python
pytest.importorskip("foo")
```

#### Docstring Test Detection:
- [x] #380

- look for patterns indicative of docstring tests (e.g., ">>>")
- or use helper functions from 3rd party modules

#### Dependency Format Detection:
- [x] #381

 (e.g., requirements.txt, requirements.in, pyproject.toml).

#### Code Directory Detection:
- [x] #382

Some heuristics:
  - Check for project name declarations in `pyproject.toml` or similar files.
  - Search for the presence of standard directories (e.g., `src/`).
  - Determine the directory with the most `.py` files.
  - Determine the directory and code structure from setup.py or pyproject.toml

```[tasklist]
### Tasks
- [ ] https://github.com/tweag/FawltyDeps/issues/377
- [ ] https://github.com/tweag/FawltyDeps/issues/378
- [ ] https://github.com/tweag/FawltyDeps/issues/379
- [ ] https://github.com/tweag/FawltyDeps/issues/380
- [ ] https://github.com/tweag/FawltyDeps/issues/381
- [ ] https://github.com/tweag/FawltyDeps/issues/382
- [ ] https://github.com/tweag/FawltyDeps/issues/398
- [ ] https://github.com/tweag/FawltyDeps/issues/402
- [ ] https://github.com/tweag/FawltyDeps/issues/403
```

**Comment by mknorps** [comment]
The work is being done on a separate branch: https://github.com/tweag/FawltyDeps/tree/zhihan/automate-PyPI-analysis

**Comment by mknorps** [comment]
Scripts inside `PyPI_analysis`,
Main difference is CLI and detect_import.py file.

**Comment by mknorps** [comment]
Possible ideas for improvements:

1. Modify the main branch to allow for custom imports parsing and in the experiment branch, add a custom parsing function
2. On the main branch, create a PR with marking different import types (regular, conditional, alternative, dynamic). That way we can extract this information from json output of FD.

**Comment by jherland** [comment]
> Possible ideas for improvements:
> 
> 1. Modify the main branch to allow for custom imports parsing and in the experiment branch, add a custom parsing function
> 2. On the main branch, create a PR with marking different import types (regular, conditional, alternative, dynamic). That way we can extract this information from json output of FD.

I believe it's fine for the PyPI analyis to remain on a separate branch of the repo. The only reason to reuse any FD code at all in the analysis is if it can speed up the progress of the analysis work. I believe that we don't want the analysis work to initially/directly affect the FawltyDeps code at all. Rather, we would run the analysis separately, and use the output from that to _then_ inform what features/changes to do in FawltyDeps?

In other words, if you want to "enrich" the JSON output from FD, that should be done on the analysis branch, so that we don't unnecessarily "lock" ourselves into providing new JSON output in released versions of FD, that then later turn out to be unnecessary/uninteresting.

Maybe I am missing an obvious benefits of a tighter integration here?

**Comment by mknorps** [comment]
The work is done on a separate branch, https://github.com/tweag/FawltyDeps/tree/zhihan/automate-PyPI-analysis.

Note: We will be adding relevant project to the real projects test suit once we identify what would be the next improvements in FawltyDeps parsing imports and dependencies.
